crypto/cms/cms_sd.c

```diff
@@ -916,9 +922,16 @@ int CMS_SignerInfo_verify(CMS_SignerInfo *si)
         goto err;
     }
     mctx = si->mctx;
+    if (si->pctx != NULL) {
+        EVP_PKEY_CTX_free(si->pctx);
+        si->pctx = NULL;
+    }
     if (EVP_DigestVerifyInit_ex(mctx, &si->pctx, EVP_MD_get0_name(md), libctx,
-                                propq, si->pkey, NULL) <= 0)
+                                propq, si->pkey, NULL) <= 0) {
+        si->pctx = NULL;
         goto err;
+    }
+    EVP_MD_CTX_set_flags(mctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX);
```

Comment:
Does `si->pctx` need to be freed or set to `NULL` after this point?
I see if being freed and then set to `NULL` on lines 926-927, set to `NULL` on line 931 without being freed, but nothing happening with `si->pctx` after line 931.

Reply:
I see the `cms_si_cb` callback frees `si->pctx`, but there's another concern below.

Reply:
yes, the callback is supposed to do that.